### PR TITLE
Mangoiv/multiple redeemer

### PR DESCRIPTION
**Pull request by MangoIV**
### What is this PR about
- add an elaborate and parametrized example that tests redeemer indexing and some other things 
- I started this while on cardax and will finish it as soon as I find time
- I was hitting some bug in ctl and I didn't have time to debug it, that's why I'm stuck. 

### WIP multipleredeemers

- I'm still hitting some issues I cannot identify the roots of. 

### Pre-review checklist

- [x] All code has been formatted using our config (`make format` for Purescript, `nixpkgs-fmt` for Nix)
- [x] All Purescript imports are explicit, including constructors
- [x] **All** existing examples have been run locally against a fully-synced testnet node
- [x] The integration and unit tests have been run (`npm run test`) locally
- [x] The changelog has been updated under the `## Unreleased` header, using the appropriate sub-headings (`### Added`, `### Removed`, `### Fixed`)

### CODE REVIEW

**Comment by rmgaray** [comment]
There's currently an issue when trying to spend a larger amount of each token. Currently the amount is set to `1` for both `TokenName`s (look at lines 103-104), the issue presents itself when increasing either of them to a larger amount. The result of this is change is a balancing phase that never ends when trying to consume all the minted tokens.

I'd appreaciate if anyone tried to reproduce the issue on their end. I've been testing this in the browser, but I assume the same thing must happen when trying the example in Plutip.

**Comment by rmgaray** [comment]
I understand that the amount of inputs of the transaction probably exceeds the maximum size, but even then CTL should fail with an error instead of hanging indefinitely

**Comment by jy14898** [comment]
@rmgaray I've had a quick look and generally this PR is good, thanks

> There's currently an issue when trying to spend a larger amount of each token.

I think this is due to setting redeemer execution units in an `n^2` manner, in `Ctl.Internal.BalanceTx.ExUnitsAndMinFee.setRdmrsExecutionUnits`. Currently we loop the execution units, and then loop the redeemers to find the match. I believe we can loop the redeemers and keep the exunits in the map, we can do it in `n*log(n)`. Thoughts @klntsky @errfrom?

**Comment by klntsky** [comment]
@jy14898 good idea! Could you please make an issue?

File: examples/MultipleRedeemers.purs [hunk 1]
```diff
@@ -0,0 +1,309 @@
...
+    configuration :: Configuration
+    configuration =
+      { validators:
+          [ red1 /\ Redeemer (toData $ Integer $ fromInt 1)
+          , red2 /\ Redeemer (toData $ Integer $ fromInt 2)
+          , red3 /\ Redeemer (toData $ Integer $ fromInt 3)
+          ]
+      , tokens:
+          [ Tuple "foo" 1
+          , Tuple "bar" 1
```
[the remark this answers is not in the document]
Comment (jy14898) [reply examples/MultipleRedeemers.purs:104]: There's a couple of places you match on the constructor in a lambda that can be changed too

File: examples/MultipleRedeemers.purs [hunk 2]
```diff
@@ -0,0 +1,309 @@
...
+          (utxo /\ red) <- utxosnreds
+          flip Constraints.mustSpendScriptOutput red <$> getorefs hash utxo
+      ]
+
+  ubTx <- liftedE $ Lookups.mkUnbalancedTx lookups constraints
+  bsTx <-
+    liftedM "Failed to balance/sign tx" $ balanceAndSignTx ubTx
+  hash2 <- submit bsTx
+  logInfo' $ "Hash of second transaction " <> show hash2
+  pure unit
```
Comment (jy14898) [inline examples/MultipleRedeemers.purs:225]:
```suggestion
  pure unit
```

File: examples/MultipleRedeemers.purs [hunk 3]
```diff
@@ -0,0 +1,309 @@
...
+  go = do
+    utxos :: Array (UtxoMap /\ Redeemer) <- for validators $ \(Tuple val red) ->
+      do
+        let vhash = validatorHash val
+        utxo <- liftContractM ("could not get utxos at " <> show vhash) =<<
+          utxosAt
+            (scriptHashAddress vhash)
+        pure $ utxo /\ red
+
+    liftAff $ delay $ Milliseconds $ toNumber 3_000
```
Comment (jy14898) [inline examples/MultipleRedeemers.purs:243]:
```suggestion
    liftAff $ delay $ Milliseconds 3_000.0
```

File: examples/MultipleRedeemers.purs [hunk 4]
```diff
@@ -0,0 +1,309 @@
...
+      ]
+
+  unbalancedTx <- liftedE $ Lookups.mkUnbalancedTx lookups constraints
+  balancedSignedTx <- liftedE $ balanceAndSignTxE unbalancedTx
+  logInfo' $ "Balanced and signed tx is " <> show balancedSignedTx
+  txId <- submit balancedSignedTx
+  logInfo' $ "Tx ID: " <> show txId
+  pure txId
+
+-- | for each Script we have one redeemer that we're goig to supply
```
Comment (jy14898) [inline examples/MultipleRedeemers.purs:190]:
```suggestion
-- | for each Script we have one redeemer that we're going to supply
```

File: examples/MultipleRedeemers.purs [hunk 5]
```diff
@@ -0,0 +1,309 @@
...
+      ]
+
+  ubTx <- liftedE $ Lookups.mkUnbalancedTx lookups constraints
+  bsTx <-
+    liftedM "Failed to balance/sign tx" $ balanceAndSignTx ubTx
+  hash2 <- submit bsTx
+  logInfo' $ "Hash of second transaction " <> show hash2
+  pure unit
+
+getUtxos
```
Comment (jy14898) [inline examples/MultipleRedeemers.purs:227]: Is there a reason why we don't `awaitTxConfirmed` the `hash` first, rather than looping until success?

**Review by jy14898 — CHANGES_REQUESTED** [review]

File: examples/MultipleRedeemers.purs (hunk 5 above)
Comment (MangoIV) [reply examples/MultipleRedeemers.purs:227]: Simply because this function didn't exist back when I wrote this.

**Comment by klntsky** [comment]
Closing, this work has been adapted in https://github.com/Plutonomicon/cardano-transaction-lib/pull/1199

**Pull request closed without merging**